WebAssets/build.js

```diff
@@ -1,40 +1,55 @@
-const path = require('path');
-const fs = require('fs');
-const jetpack = require('fs-jetpack');
-const { task, tasks, run } = require('oldowan');
-
-task('js', async () => {
-    const structure = await jetpack.readAsync('./js/structure.js');
-    const final = structure.replace(/\s*\/\/ include:\s*(\S+)\s*/g, (m, $1) => {
-        const includePath = path.resolve(path.join('./js', $1));
-        return '\r\n\r\n    ' + jetpack.read(includePath)
-            .replace(/\/\* (?:globals|exported).+\*\//g, '')
-            .trim()
-            .replace(/\n/g, '\n    ');
-    });
-    await jetpack.writeAsync('dist/mirrorsharp.js', final);
-}, { inputs: ['js/*.js'] });
-
-task('d.ts', () => jetpack.copyAsync('types/public.d.ts', 'dist/mirrorsharp.d.ts', { overwrite: true }), { inputs: ['types/public.d.ts'] });
+import jetpack from 'fs-jetpack';
+import fg from 'fast-glob';
+import ts from 'typescript';
+import oldowan from 'oldowan';
+const { task, tasks, run } = oldowan;
+
+task('ts', async () => {
+    const { options } = ts.getParsedCommandLineOfConfigFile('ts/tsconfig.json', undefined, ts.sys);
+    const program = ts.createProgram(['ts/mirrorsharp.ts'], Object.assign(options, {
+        module: ts.ModuleKind.ES2015,
+        noEmit: false,
+        outDir: 'dist',
+        declaration: true
+    }));
+
+    const emitResult = program.emit();
+    const diagnostics = ts
+        .getPreEmitDiagnostics(program)
+        .concat(emitResult.diagnostics);
+
+    for (const diagnostic of diagnostics) {
+        if (diagnostic.file) {
+            const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
+            const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n");
+            console.log(`${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`);
+        } else {
+            console.log(ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n"));
+        }
+    }
+    if (emitResult.emitSkipped)
+        throw new Error("TypeScript compilation failed.");
+
+    // Add .js extension to all imports.
+    // Technically TypeScript already resolves .js to .ts, but it's a hack.
+    await Promise.all((await fg(['dist/**/*.js'])).map(async path => {
+        const content = await jetpack.readAsync(path);
+        const replaced = content.replace(/from '(\.[^']+)';/g, "from '$1.js';");
+        await jetpack.writeAsync(path, replaced);

```

Comment:
Instead of `async await` here, why not just return `writeAsync` promise

Reply:
Seems more consistent / easier to follow? Also not much performance difference given it's the build.